zebra-consensus/src/checkpoint.rs

```diff
@@ -754,16 +795,74 @@ impl Service<Arc<Block>> for CheckpointVerifier {
         // on the next call(). Failures always reject a block, so we know
         // there will be at least one more call().
         //
-        // TODO(teor): retry on failure (low priority, failures should be rare)
-        self.process_checkpoint_range();
+        // We don't retry with a smaller range on failure, because failures
+        // should be rare.
+        let valid_blocks = self.process_checkpoint_range();
 
         metrics::gauge!("checkpoint.queued_slots", self.queued.len() as i64);
 
-        async move {
-            // Remove the Result<..., RecvError> wrapper from the channel future
-            rx.await
-                .expect("CheckpointVerifier does not leave dangling receivers")
+        // Optimisation: return a small future in the common case.
+        //
+        // Rust 1.46 and later have known regressions in future size and
+        // complexity, so we want to return a small future in the common case.
+        if valid_blocks.is_empty() {
+            async move {
+                // Remove the Result<..., RecvError> wrapper from the channel future.
+                rx.await
+                    .expect("CheckpointVerifier does not leave dangling receivers")
+            }
+            .boxed()
+        } else {
+            let mut state_service = self.state_service.clone();
+
+            async move {
+                // All the blocks in the current range are valid. Commit them to the
+                // state.
+                for qblock in valid_blocks {
+                    // Since we're executing in a future, these blocks could be
+                    // committed out of order. That's ok, the state service will
+                    // handle out-of-order blocks.
+                    let ready_state = state_service.ready_and().await?;
+
+                    let result = match ready_state
+                        .call(zebra_state::Request::AddBlock {
+                            block: qblock.block.clone(),
+                        })
+                        .await?
+                    {
+                        zebra_state::Response::Added {
+                            hash: committed_hash,
+                        } => {
+                            assert_eq!(
+                                committed_hash, qblock.hash,
+                                "state returned wrong hash: hashes must be equal"
+                            );
+                            Ok(qblock.hash)
+                        }
+                        _ => Err(format!(
+                            "adding block {:?} {:?} to state failed",
+                            qblock.block.coinbase_height(),
+                            qblock.hash
+                        )
+                        .into()),
+                    };
+
+                    // Finally, tell the caller the result.
+                    // Sending can fail, but there's nothing we can do about it.
+                    let _ = qblock.tx.send(result);
```

Comment:
I'm not sure this logic makes sense... the `result` that's sent to the caller is not the result of that caller's block, but the result of some other blocks.

But because we're allowed to commit blocks to the state out of order, I don't think it's necessary to have two codepaths.  The old code did
```rust
        async move {
            // Remove the Result<..., RecvError> wrapper from the channel future
            rx.await
                .expect("CheckpointVerifier does not leave dangling receivers")
        }
        .boxed()
```
and the new code can commit the block after it learns it's ready:
```rust
        async move {
            match rx
                .await
                .expect("CheckpointVerifier does not leave dangling receivers")
            {
                Ok(hash) => {
                    state_service
                        .oneshot(zebra_state::Request::AddBlock { block })
                        .await?;
                    Ok(hash)
                }
                Err(e) => Err(e),
            }
        }
        .boxed()
```

Reply:
> I'm not sure this logic makes sense... the `result` that's sent to the caller is not the result of that caller's block, but the result of some other blocks.

Can you explain what you mean here?

As far as I can see, each `result` is based on a `qblock.block`, and it is sent to the corresponding `qblock.tx`.

The `rx` later in the async block is totally unrelated, although it might match one of the earlier `qblock.tx` channels.

But I can see an issue with error handling here, where the `?` operator takes errors from any qblock, and propagates them to the unrelated current block.

So I'll implement this change to fix that issue.

Reply:
Implemented and squashed into f9088027f207577b08b0a7450d1ce05bc2e31c54.

This change commits blocks in an even more unordered fashion, so I needed to ignore another test. Squashed into c042555ac714819b03efbc565bc0223fedf43102.